solr/core/src/java/org/apache/solr/cluster/placement/plugins/AffinityPlacementFactory.java

```diff
@@ -555,11 +579,14 @@ public Map<Replica, String> canRemoveReplicas(Collection<Replica> replicas) {
                       replica.getShard().getCollection().getName(), k -> new HashMap<>())
                   .computeIfAbsent(replica.getShard().getShardName(), k -> new HashSet<>());
           replicasRemovedForShard.add(replica);
-
-          if (replicasRemovedForShard.size()
-              >= getReplicasForShardOnNode(replica.getShard()).size()) {
-            replicaRemovalExceptions.put(
-                replica, "co-located with replicas of " + collocatedCollections);
+          // either if all shards are mandatory, or the current one is mandatory
+          if (mandatoryShardsOrAll.isEmpty()
+              || mandatoryShardsOrAll.contains(replica.getShard().getShardName())) {
+            if (replicasRemovedForShard.size()
+                >= getReplicasForShardOnNode(replica.getShard()).size()) {
+              replicaRemovalExceptions.put(
```

Comment:
This exception should likely mention the shard that is affected.

Reply:
ok. I'll add shard name, but replica name already contains a shard name, I suppose.